Use Towncrier to generate the unstable changelog

turt2live opened the pull request:
Why we should use Towncrier:
* Towncrier makes it nearly impossible to have a merge conflict on the changelog
* Generating changelogs is so much easier

This still requires a contributor to open a pull request before they can add to the changelog, however that seems to have worked okay up until now and there's not really a way around it.

A rendered version of the spec is available here: ***TODO***

## Notable changes
* The changelog format is different now (towncrier is a little opinionated about the format, but it is at least clean)
* The project now uses Python 3, which means ***everybody*** needs a new virtual environment, including any automated builds.
* If you haven't guessed by now, [Towncrier](https://github.com/hawkowl/towncrier) is used for changelogs

richvdh commented on scripts/gendoc.py:
```diff
@@ -395,15 +397,16 @@ def get_group(group_id, depth):
 
 def cleanup_env():
-    shutil.rmtree(tmp_dir)
+    #shutil.rmtree(tmp_dir)
```
this looks bogus

turt2live replied:
yea, I noticed that too - my bad.

richvdh commented on scripts/templating/matrix_templates/units.py:
```diff
@@ -831,12 +833,28 @@ def load_changelogs(self):
+                try:
+                    raw_log = subprocess.check_output(
+                        ['towncrier', '--version', 'Unreleased Changes', '--name', name, '--draft'],
+                        stderr=subprocess.PIPE,
```
if this fails, will the output just get swallowed with no clue as to what happened?

richvdh replied:
I'm sorry, you're saying that towncrier sends the changelog output to stderr and error/progress messages to stdout?! That's... insane.

Either way, I think the best thing would be to print all the output (both stdout and stderr) if the command exits non-zero.

turt2live replied:
progress goes to stdout, changelog+errors go to stderr. In any case: yea - I'll just make it print a bunch of stuff to the log if it blows up.

turt2live replied:
looks like I'm actually just dumb and don't understand how the subprocess stuff works. Towncrier isn't insane and sends the log via stdout, with errors+progress over at stderr.

turt2live commented on api/check_examples.py:
```diff
@@ -88,9 +88,9 @@ def check_swagger_file(filepath):
         swagger = yaml.load(f)
 
-    for path, path_api in swagger.get('paths', {}).items():
+    for path, path_api in list(swagger.get('paths', {}).items()):
```
python's 2to3 did that :( I'm not entirely sure why - it felt like something not worth fighting at the time, although I can go through and fix things if needed.

turt2live replied:
I'll take up the fight against python and remove them.

turt2live commented:
A bunch of lists have been removed, and potentially improvements in the area of towncrier exploding on us. @richvdh please take a look when you get a chance.

richvdh commented on scripts/templating/matrix_templates/units.py:
```diff
@@ -29,8 +29,10 @@
 import sys
-import urllib
 import yaml
+import requests.utils
```
I'm not keen on adding a dep on `requests` just to get access to a uri escaping function. What's wrong with [urllib.parse.quote](https://docs.python.org/3.7/library/urllib.parse.html#urllib.parse.quote) ? Or, equivalently, since you're using six anyway, `six.moves.urllib.parse.quote` ?

turt2live replied:
`requests` is used by the proposals script, so I thought I'd use it here too. Using six probably makes more sense

richvdh commented on scripts/templating/matrix_templates/units.py:
```diff
@@ -831,12 +833,42 @@ def load_changelogs(self):
+            tc_lines = []
+            if os.path.isdir(tc_path):
+                logger.info("Generating towncrier changelog for: %s" % name)
+                p = subprocess.run(
```
subprocess.run was added in python 3.5; our speculator runs on a box with debian jessie which only has 3.4. We could upgrade it, but... could you use `subprocess.call` instead?

turt2live commented:
Should now be python 3.4 compatible

hawkowl commented on scripts/requirements.txt:
```diff
@@ -7,3 +7,5 @@ Jinja2 >= 2.9.6
 jsonschema >= 2.6.0
 PyYAML >= 3.12
 requests >= 2.18.4
+towncrier == 18.6.0rc1
```
18.6.0 final is out now :)

hawkowl commented:
FWIW, the porting looks reasonable. I don't love towncrier being used through popen though, but that's on me for not providing a scripting based API :)

If you want to change the changelog format, you can provide it another template, although I believe towncrier's default is probably worth keeping for its uniformity.

turt2live commented:
A scripting API would be a big +1 from me :)  The default format (within reason) is useful for the actual changelog we generate at release-time, so the build process gets to be the exception and strip out headers.

turt2live commented:
@hawkowl if you're comfortable with it, an approval on this would be great btw :)

richvdh commented on README.rst:
```diff
@@ -41,7 +41,7 @@ specs and event schemas in this repository.
 -----------
 
-To use the scripts, it is best to create a Python 2.x virtualenv as follows::
+To use the scripts, it is best to create a Python 3.5+ virtualenv as follows::
```
presumably the nix stuff is now outdated. We should probably remove it.

richvdh commented on README.rst:
```diff
@@ -41,7 +41,7 @@ specs and event schemas in this repository.
-To use the scripts, it is best to create a Python 2.x virtualenv as follows::
+To use the scripts, it is best to create a Python 3.5+ virtualenv as follows::
 
   virtualenv env
```
`virtualenv -p python3 env`

richvdh commented on .travis.yml:
```diff
@@ -2,15 +2,16 @@ language: go
+install:
+- sudo apt-get update
+- sudo apt-get install python3.5 python3.5-dev
+- sudo pip install --upgrade virtualenv
```
is this necessary? If it is, it should be in the README, but upgrading distro stuff with pip is bad karma.

richvdh commented on scripts/test-and-build.sh:
```diff
@@ -4,8 +4,13 @@ set -ex
 cd `dirname $0`/..
 
-virtualenv env
+virtualenv -p python3.5 env
```
can we make this `-p python3` so that it will work on systems that have 3.6 or later?

turt2live commented:
Have addressed your review: the tldr is python 3.4 works with matrix-doc, so we don't need any of the python 3.5 stuff that was geared towards Travis CI.
